std-bits/table/src/main/java/org/enso/table/excel/ExcelRow.java

```diff
@@ -40,14 +39,24 @@ public Object getCellValue(int column) {
         double dblValue = cell.getNumericCellValue();
         if (DateUtil.isCellDateFormatted(cell) && DateUtil.isValidExcelDate(dblValue)) {
           var dateTime = DateUtil.getLocalDateTime(dblValue);
-          if (dateTime.isBefore(LocalDateTime.of(1900, 1, 2, 0, 0))) {
-            // Excel stores times as if they are on the 1st January 1900.
-            // Due to the 1900 leap year bug might be 31st December 1899.
-            return dateTime.toLocalTime();
-          }
-          if (dateTime.getHour() == 0 && dateTime.getMinute() == 0 && dateTime.getSecond() == 0) {
+          var dateFormat = cell.getCellStyle().getDataFormatString();
+          var hasTime =
+              dateFormat.contains("h")
+                  || dateFormat.contains("H")
+                  || dateTime.getHour() != 0
+                  || dateTime.getMinute() != 0
+                  || dateTime.getSecond() != 0;
+          var hasDate =
+              dateFormat.contains("d")
+                  || dateFormat.contains("D")
+                  || dateFormat.contains("y")
+                  || dateFormat.contains("Y");
+          if (hasDate && !hasTime) {
             return dateTime.toLocalDate();
           }
+          if (!hasDate && hasTime) {
+            return dateTime.toLocalTime();
+          }
```

Comment:
```suggestion
          if (dateTime.isBefore(LocalDateTime.of(1900, 1, 2, 0, 0))) {
            // Excel stores times as if they are on the 1st January 1900.
            // Due to the 1900 leap year bug might be 31st December 1899.
            return dateTime.toLocalTime();
          }
          if (dateTime.getHour() == 0 && dateTime.getMinute() == 0 && dateTime.getSecond() == 0) {
            var dateFormat = cell.getCellStyle().getDataFormatString();
            if (!dateFormat.contains("h") && !dateFormat.contains("H")) {
              return dateTime.toLocalDate();
            }
          }
          return dateTime.atZone(ZoneId.systemDefault());
```

For the minimal change?
The 1900 fix is jibberish so we should fix that.

Reply:
Sure. If it makes my test data round trip successfully and all the other tests pass then it works for me. Both versions feel ugly.

Reply:
Done